### ISSUE TITLE
[iOS][WebView] Gesture recognizers not blocked in iOS 13.4 and 13.5

### ISSUE THREAD

**Issue by iKK001** [flutter/flutter#53490]
After having installed Mac OS Catalina 10.15.4, Xcode 11.4 and iOS13.4, my Flutter Application running on iOS (simulator and actual device [iPhoneX]) is very buggy !

Before, I had a `floatingActionButton` working fine. And also `Drawer` (left and right) where working fine.

Underneath the FloatingActionButton and Drawers is a WebView.

But after the Mac SW update:
- `floatingActionButton` makes a screen-freeze
- `Drawer` shows but freezes when trying to close it

I tried the following two configs:
1.) `Flutter STABLE 1.12.13+hotfix.8`,  plus the Xcode-11.4 workaround described [here...](https://flutter.dev/docs/development/ios-project-migration)
2.) `Flutter BETA 1.15.17`

But, again, for both configs same error behaviour under iOS - screen-freezes without error-message.

On Android, everything works fine.

What is wrong under iOS?

Here excerpts of the `pubspec.yaml` :

```dart
environment:
  sdk: ">=2.1.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  flutter_localizations:
    sdk: flutter
  webview_flutter: ^0.3.19+9
  cloud_firestore: ^0.13.4+1
  firebase_core: ^0.4.4+2
  firebase_auth: 0.15.3
  provider: ^4.0.4
  flutter_spinkit: ^4.1.2
  device_info: ^0.4.2+1
  page_transition: ^1.1.5
  package_info: ^0.4.0+16
  crypto: ^2.1.3
  shared_preferences: ^0.5.6+3
  url_launcher: ^5.4.2
  flutter_background_geolocation: ^1.7.1

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^0.1.2

dependency_overrides:
  firebase_core: 0.4.4

dev_dependencies:
  flutter_test:
    sdk: flutter
```

Here is my `flutter doctor -v` :

```dart
[✓] Flutter (Channel beta, v1.15.17, on Mac OS X 10.15.4 19E266, locale en-CH)
    • Flutter version 1.15.17 at /Users/myName/Documents/flutter
    • Framework revision 2294d75bfa (3 weeks ago), 2020-03-07 00:28:38 +0900
    • Engine revision 5aff311948
    • Dart version 2.8.0 (build 2.8.0-dev.12.0 9983424a3c)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/myName/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4, Build version 11E146
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.43.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.8.1

[✓] Connected device (2 available)
    • Android SDK built for x86 • emulator-5554                            • android-x86 • Android 10 (API 29) (emulator)
    • MyPhone                     • d709a13432kjie7429258466dbas3f29r45gdl8s • ios         • iOS 13.4

• No issues found!
```

edited by cyanglaz
Internal: b/155322611

**Comment by VladyslavBondarenko** [comment]
Hi @iKK001 
Can you post complete minimal code sample to reproduce your case?
With this simplest code I don't experience any freezes.
Tried with `beta 1.15.17` and iPhone 7 iOS 13.4 after MacOS and Xcode update

<details>
 <summary>code</summary>

```
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  final appTitle = 'Drawer Demo';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: appTitle,
      home: MyHomePage(title: appTitle),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  MyHomePage({Key key, this.title}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: WebView(
        initialUrl: 'https://flutter.dev/',
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          print('pressed');
        },
        child: Icon(Icons.navigation),
        backgroundColor: Colors.green,
      ),
      drawer: Drawer(
        child: ListView(
          padding: EdgeInsets.zero,
          children: <Widget>[
            DrawerHeader(
              child: Text('Drawer Header'),
              decoration: BoxDecoration(
                color: Colors.blue,
              ),
            ),
            ListTile(
              title: Text('Item 1'),
              onTap: () {
                Navigator.pop(context);
              },
            ),
            ListTile(
              title: Text('Item 2'),
              onTap: () {
                Navigator.pop(context);
              },
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

https://github.com/flutter/flutter/issues/53510, https://github.com/flutter/flutter/issues/53511

**Comment by iKK001** [comment]
Hi Vladyslav, here is my code (i.e. most of it to get the idea):

```dart
import 'dart:async';
import 'package:myapp_test/app_locations.dart';
import 'package:myapp_test/drawers/drawer_left.dart';
import 'package:myapp_test/drawers/drawer_right.dart';
import 'package:flutter/material.dart';
import 'dart:convert';
import 'package:myapp_test/shared/constants.dart';
import 'package:webview_flutter/webview_flutter.dart';
import 'package:myapp_test/screens/user_state.dart';

final webViewKey = GlobalKey<_WebViewContainerState>();

class MyWebView extends StatefulWidget {
  @override
  _MyWebViewState createState() => _MyWebViewState();
}

class _MyWebViewState extends State<MyWebView> {
  String currentPage = 'page001';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: Builder(
          builder: (context) => IconButton(
              icon: Icon(Icons.menu), // Icon(Icons.more_vert),
              onPressed: () {
                Scaffold.of(context).openDrawer();
              }),
        ),
        title: const Text(
          'My App',
          style: TextStyle(fontSize: 23.0),
        ),
        // This drop down menu demonstrates that Flutter widgets can be shown over the web view.
        actions: <Widget>[
          // IconButton(icon: Icon(Icons.search), onPressed: () async { },),
          Builder(
            builder: (context) => GestureDetector(
              onTap: () {
                Scaffold.of(context).openEndDrawer();
              },
              child: Container(
                width: 60.0,
                height: 44.0,
                child: Center(
                  child: Image.asset(
                    'assets/icons/page.png',
                    fit: BoxFit.fill,
                    height: 34.0,
                    width: 34.0,
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
      drawer: SafeArea(
        child: Transform.translate(
          offset: Offset(0.0, -50.0),
          child: ClipRRect(
            borderRadius: BorderRadius.only(
              topLeft: Radius.circular(0.0),
              topRight: Radius.circular(14.0),
              bottomLeft: Radius.circular(0.0),
              bottomRight: Radius.circular(14.0),
            ),
            child: Container(
              height: 620.0,
              child: Drawer(
                elevation: 4.0,
                child: DrawerLeft(),
              ),
            ),
          ),
        ),
      ),
      endDrawer: SafeArea(
        child: Transform.translate(
          offset: Offset(0.0, -50.0),
          child: ClipRRect(
            borderRadius: BorderRadius.only(
              topLeft: Radius.circular(14.0),
              topRight: Radius.circular(0.0),
              bottomLeft: Radius.circular(14.0),
              bottomRight: Radius.circular(0.0),
            ),
            child: Container(
              height: 620.0,
              child: Drawer(
                elevation: 4.0,
                child: DrawerRight(
                  callback: (pageType) {
                    setState(() {
                      currentPage = pageType;
                      // using currentState with question mark to ensure it's not null
                      webViewKey.currentState?.reloadWebView(pageType);
                    });
                  },
                ),
              ),
            ),
          ),
        ),
      ),
      body: WebViewContainer(key: webViewKey),
      floatingActionButton: favoriteButton(webViewKey),
    );
  }

  Widget favoriteButton(GlobalKey<_WebViewContainerState> webViewKey) {
    return Container(
      width: 500.0,
      height: 80.0,
      child: Stack(
        children: <Widget>[
          Positioned(
            bottom: 10.0,
            left: 30.0,
            child: ButtonTheme(
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(15.0)),
              height: 60.0,
              minWidth: 345.0,
              child: RaisedButton(
                child: Text(
                  AppLocalizations.of(context)
                      .translate('    My Personal Situation'),
                  style: TextStyle(
                      fontSize: 20.0,
                      color: Colors.white,
                      fontWeight: FontWeight.w600),
                ),
                color: Colors.blue,
                onPressed: () async {
                  return await Navigator.push(
                    context,
                    MaterialPageRoute(builder: (context) => UserState()),
                  ).then((value) {
                    setState(() {
                      webViewKey.currentState?.reloadWebView(currentPage);
                    });
                  });
                },
              ),
            ),
          ),
          Positioned(
            bottom: 18.0,
            left: 40.0,
            child: ClipRRect(
              borderRadius: BorderRadius.circular(8.0),
              child: Image.asset(
                'assets/icons/icon.png',
                fit: BoxFit.cover,
                height: 42.0,
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class WebViewContainer extends StatefulWidget {
  WebViewContainer({Key key}) : super(key: key);

  @override
  _WebViewContainerState createState() => _WebViewContainerState();
}

class _WebViewContainerState extends State<WebViewContainer> {
  final Completer<WebViewController> _webViewController =
      Completer<WebViewController>();
  String myWebSide = '';

  String _getPage_001(
      String pageName, String lat, String lng, String zoom) {
    return '''
    <!DOCTYPE html>
    <html>
      <head>
        <meta name="viewport" content="width=device-width, initial-scale=1.0">

        <style>
          /* Always set the map height explicitly to define the size of the div
          * element that contains the map. */
          #map {
            height: 100%;
          }
          /* Optional: Makes the sample page fill the window. */
          html, body {
            height: 100%;
            margin: 0;
            padding: 0;
          }          
        </style>
      </head>
      <body>

        <div id="map"></div>

        <script>
          const maxI = 1.0, rad = 27.0, opac = 0.6; weight = 1.0; dissipating = true;
          // const maxI = 1.0, rad = 0.01, opac = 0.6; weight = 0.5, dissipating = false;
          var map;
          var darkmode = true;
          var darkmodeStyle = [];
          var zoom = $zoom;
          var centerPos = {
            lat: $lat,
            lng: $lng
          };
          var mapOptions = {
                zoom: zoom,
                // maxZoom: 15,
                // minZoom: 12,
                center: {lat: centerPos.lat, lng: centerPos.lng},
                mapTypeId: 'roadmap',
                styles: lightStyles,
                fullscreenControl: false,
                streetViewControl: false,
                scaleControl: false,
                zoomControl: false,
          };

          function initMap() {

            // console.log(centerPos.lat, centerPos.long);
            map = new google.maps.Map(document.getElementById('map'), mapOptions);
          }

        </script>
        <script async defer
            src="https://maps.googleapis.com/maps/api/js?key=${APIKeys.googleMapsKey}&libraries=visualization&callback=initMap">
        </script>
      </body>
    </html>
    ''';
  }

  @override
  Widget build(BuildContext context) {

    return Builder(builder: (BuildContext context) {
      String contentBase64 =
          base64Encode(const Utf8Encoder().convert(myWebSide));
      return WebView(
        initialUrl: 'data:text/html;base64,$contentBase64',
        javascriptMode: JavascriptMode.unrestricted,
        onWebViewCreated: (WebViewController webViewController) {
          _webViewController.complete(webViewController);          
        },
        javascriptChannels: <JavascriptChannel>[
          _toasterJavascriptChannel(context),
        ].toSet(),
        navigationDelegate: (NavigationRequest request) {
          if (request.url.startsWith('https://www.youtube.com/')) {
            // print('blocking navigation to $request}');
            return NavigationDecision.prevent;
          }
          // print('allowing navigation to $request');
          return NavigationDecision.navigate;
        },
        onPageStarted: (String url) {
          // print('Page started loading: $url');
        },
        onPageFinished: (String url) {
          // print('Page finished loading: $url');
        },
        gestureNavigationEnabled: true,
      );
    });
  }

  void reloadWebView(String pageType) async {
    if (pageType == 'page001') {
      String pageName = 'Page001.png';
      String lat = '46.696605';
      String lng = '8.294296';
      String zoom = '7';
      myWebSide = _getPage_001(pageName, lat, lng, zoom);
    }
    // reload new webview
    setState(() {
      String contentBase64 =
          base64Encode(const Utf8Encoder().convert(myWebSide));
      _webViewController.future.then((value) => value.loadUrl('data:text/html;base64,$contentBase64'));
    });
  }
}
```

**Comment by VladyslavBondarenko** [comment]
@iKK001 
could you record your screen to show how does it look?

**Comment by iKK001** [comment]
Below you find a video that illustrates the error behaviour:
- All gestures and touch-event are "transparent" and make the webview underneath move instead.
- All navigation events are ignored
- The only working-area is the NavBar (i.e. where there is no webView)
- If I replace the webview by a simple Text('hello') Widget instead, then everything works again...
  (problem only for iOS)

[webview_error.mp4.zip](https://github.com/flutter/flutter/files/4402682/webview_error.mp4.zip)

**Comment by xuexin** [comment]
Any new information on this issue？

**Comment by DFelten** [comment]
Same problem here. On our [WebView App](https://apps.apple.com/de/app/camping-info-campingf%C3%BChrer/id1138357341) there is a map and when selecting a point of interest, then close it, the WebView is freezed. The back button in the navigation bar is still working.

**Comment by iKK001** [comment]
Vaclav, do you have any comment on this, please ?  Is it a Flutter issue or not ?

**Comment by VladyslavBondarenko** [comment]
@iKK001 
sorry, I don't have comments about this. But I just realised that it is like this when using google map in WebView. Reproduces with `dev v1.16.4-pre.33` as well

<details>
 <summary>code</summary>

```
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  final appTitle = 'Drawer Demo';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: appTitle,
      home: MyHomePage(title: appTitle),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  MyHomePage({Key key, this.title}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: WebView(
        initialUrl: 'https://www.google.ee/maps',
        javascriptMode: JavascriptMode.unrestricted,
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          print('pressed');
        },
        child: Icon(Icons.navigation),
        backgroundColor: Colors.green,
      ),
      drawer: Drawer(
        child: ListView(
          padding: EdgeInsets.zero,
          children: <Widget>[
            DrawerHeader(
              child: Text('Drawer Header'),
              decoration: BoxDecoration(
                color: Colors.blue,
              ),
            ),
            ListTile(
              title: Text('Item 1'),
              onTap: () {
                Navigator.pop(context);
              },
            ),
            ListTile(
              title: Text('Item 2'),
              onTap: () {
                Navigator.pop(context);
              },
            ),
          ],
        ),
      ),
    );
  }
}
``` 

</details>

1) without using floatingActionButton, open drawer, now map responds to gestures instead of drawer's items.
2) drag map, press floatingActionButton. After this map doesn't respond to gestures, but drawer's items do.

<details>
 <summary>Screen recording</summary>

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/17024977/78231265-34294080-74db-11ea-8c1e-1677abf0463c.gif)

</details>

**Comment by DFelten** [comment]
@VladyslavBondarenko The error also occurs with HereMaps. We use HereMaps on the page that is integrated in the WebView. 

The error occurs on this page when it is integrated in a WebView: https://www.camping.info/en/search-on-map?area=-21.240625,61.111082,42.040624,38.358364&zl=5

**Comment by xuexin** [comment]
@VladyslavBondarenko This issue is also occurs with my inline web page(using webview_flutter), I think title is not correct now.

**Comment by VladyslavBondarenko** [comment]
I’m not sure what severity label is better to use here, please consider to change it

**Comment by cyanglaz** [comment]
@VladyslavBondarenko plugin issues usually doesn't block dev rolls, I'm going to lower the severity.

**Comment by tvolkert** [comment]
@VladyslavBondarenko following on @cyanglaz, this is because plugins are published and versioned independently from the framework (though they are often dependent on framework versions/features).  So a plugin bug will rarely cause us to stop publishing framework releases (unless the bug is affecting a plugin but actually lies in the framework).

**Comment by cyanglaz** [comment]
To summarize my discovery:
Confirmed that it only started to happen on iOS 13.4, and it only affects certain web pages. ("google map" webpage has the issue, but "youtube" works fine)

There are 2 issues here:
1. The gesture is recognized by the webview behind the flutter button.
The framework successfully recognized the gesture tapped on the button, and tries to block the webview to receive the guesture. However, the webview continues to receive the gesture. Which is similar to https://github.com/flutter/flutter/issues/50420; And it is probably due to some changes in the WKWebView that they started to override UIGestureRecognizer methods like [touchesBegin:](https://developer.apple.com/documentation/uikit/uiresponder/1621142-touchesbegan?language=objc) to handle gestures. We have a known issue that our gesture block implementation on iOS doesn't work if the platform view implements "onTouchesBegan" and "onTouchesEnded" to handle gestures, rather than use the standard [addTarget:action:](https://developer.apple.com/documentation/uikit/uigesturerecognizer/1624230-addtarget?language=objc) method.
2. Our [ForwardingGestureRecognizer](https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm#L691) didn't receive [touchesEnded:](https://developer.apple.com/documentation/uikit/uigesturerecognizer/1620005-touchesended?language=objc) callback. Then flutter framework didn't receive the touches up event, resulting the `onTap` not getting called. 

Both issues are newly introduced in iOS 13.4, and I suspect both are due to some changes in WKWebView.
Will give further update in this thread if when I have any progress.

**Comment by cyanglaz** [comment]
cc @amirh

**Comment by abellin** [comment]
I am also experiencing this as of updating to Catalina/Xcode 11.4 tonight.

**Comment by Sanjys** [comment]
@cyanglaz is there any progress，please？We are also waiting for the solution to this problem

**Comment by Nico04** [comment]
Same issue here, using a WebView with a 360° panorama viewer.
Issue appeared on iOS 13.4, works fine on Android.

**Comment by cyanglaz** [comment]
https://stackoverflow.com/questions/61208881/cant-handle-touches-simultaneously-with-wkwebview-on-ios-13-4 
seems related. It seems like WKWebView starts to block other gesture recognizers that are added to the view.
We use [DelayingGestureRecognizer](https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm#L695) and [ForwardingGestureRecognizer](https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm#L720) on the WKWebview to handle gestures in flutter, and both of them are blocked by WKWebView. This caused discrepancy in the gesture handling in Flutter.

**Comment by DFelten** [comment]
> @VladyslavBondarenko The error also occurs with HereMaps. We use HereMaps on the page that is integrated in the WebView.
> 
> The error occurs on this page when it is integrated in a WebView on iOS 13.4: https://www.camping.info/en/search-on-map

We have now switched to the [Community Plugin for WebViews](https://pub.dev/packages/flutter_webview_plugin) because the map works there. This is maybe an alternative or an interim solution for some.

**Comment by Nico04** [comment]
> We have now switched to the [Community Plugin for WebViews](https://pub.dev/packages/flutter_webview_plugin) because the map works there. This is maybe an alternative or an interim solution for some.

A workaround to this issue is to remove all Flutter Widgets that are shown above the WebView, which is very inconvenient.
But your suggestion to use the Community plugin is the same no ?
From the main page of the latter :
_Warning: The webview is not integrated in the widget tree, it is a native view on top of the flutter view. You won't be able see snackbars, dialogs, or other flutter widgets that would overlap with the region of the screen taken up by the webview._

**Comment by DFelten** [comment]
For our purpose it's an alternative, because we're using currently just a pure WebView app without any overlay widgets. So the app can at least be used again under iOS 13.4.

**Comment by Nico04** [comment]
> For our purpose it's an alternative, because we're using currently just a pure WebView app without any overlay widgets. So the app can at least be used again under iOS 13.4.

So you have this issue even without Flutter Widgets above the WebView ?

**Comment by DFelten** [comment]
> So you have this issue even without Flutter Widgets above the WebView ?

Yes, like I said above we've got an app with only a WebView. It's possible to reproduce the freezing of the WebView in the AppStore app or just using a WebView with this url (https://www.camping.info/en/search-on-map). Then the following must be done:

> Same problem here. On our [WebView App](https://apps.apple.com/de/app/camping-info-campingf%C3%BChrer/id1138357341) there is a map and when selecting a point of interest, then close it (go back to the map), the WebView is freezed. The back button in the navigation bar is still working.

**Comment by cyanglaz** [comment]
@DFelten Your issue might be a little different. If you could open a new issue with reproduction details and reduced test case, we can surely take a look!

**Comment by Smagold** [comment]
any updates?

**Comment by lboel** [comment]
Any workaround so far? 
The issue seems to persist in the iOS 13.4.5beta. Isn't a stable webview gesture support a major issue for many developers? Maybe someone can share his or her approach to deal with the problem.

**Comment by iapicca** [comment]
Hi @iKK001 
are you still experiencing the issue with the latest master?
[sample](https://github.com/flutter/flutter/issues/56426#issuecomment-624514639)
thank you

**Comment by iKK001** [comment]
Sorry, I could not test (since very busy right now). I quick-tried to let my project run under master (`1.18.0-11.0.pre.6 • channel master •` - but unfortunately, a new problem occurred `The method 'RadioSliderThumbShape.paint' has fewer named arguments than those of overridden method 'SliderComponentShape.paint'` - and I really do not have time to investigate this right now. As long as Flutter changes versions with breaking changes, it is not useful for a production App.

**Comment by macoshita** [comment]
@cyanglaz I'm facing a slightly different issue than this one, but I believe the cause is the same.
When this happens, the value of `DelayingGestureRecognizer.state` seems to be unchanged.

The following is my guess. I'm sorry if I'm wrong. 🙇 

I'm curious about [this code](https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm#L839).
If `shouldEndInNextTouchesEnded` is set to `YES`, then `state` is set to `UIGEstureRecognizerStateEnded`.
But it is only when `touchesEnded` is fired. No action is taken when `touchesCancelled` is fired. 
If `touchesCancelled` is fired, is `state` set to `UIGeStureRecognizerStateFailed` ?

**Comment by macoshita** [comment]
I conducted the following experiment.

- I got the `DelayingGestureRecognizer` and `ForwardingGestureRecognizer` from the `UIView#gestureRecognizers` and monitored the values.
- I noticed that `ForwardingGestureRecognizer.state` sometimes remains `possible` even if `DelayingGestureRecognizer.state` becomes `failed`.
- Then I set `failed` to `ForwardingGestureRecognizer.state`, and the bug was resolved.

**Comment by cyanglaz** [comment]
@macoshita Nice! Could you explain more? Or if you'd like to start a PR I'd be happy to take a look!

**Comment by andrearubeis** [comment]
@macoshita thank you for your debug ! Could you explain more about your solution ? 
I have a production app that with this issue is almost unusable.

**Comment by macoshita** [comment]
One moment, please. I'll make a small reproduction example!

**Comment by macoshita** [comment]
[My experimental code](https://github.com/macoshita/plugins/commit/65113dbf3180f972ab7f0ed60fe35f545ec65b9f)

This is a webview_flutter example with a "Fix Freeze" button.
If webview no longer responds to gesture, pressing this button will make it respond again.
If you press this button when webview is unresponsive, debug console will be shown as `recognizer delay_state: 5, forward_state: 0`.
Whenever this problem occurs, [`UIGestureRecognizer.state`](https://developer.apple.com/documentation/uikit/uigesturerecognizer/state) is different from `ForwardingGestureRecognizer` and `DelayingGestureRecognizer`.
In most cases, "Delaying..." is 5(meaning: failed) and "Forwarding..." is 0(meaning: possible). In rare cases, "Delaying..." is 3(meaning: ended).
If observing this state and make it automatically rewrite the state, this problem should be fixed.
However, this is not a radical solution at all.

Please see also this reproduction video (iPhone 8 Plus)↓

![fixfreeze](https://user-images.githubusercontent.com/6210798/82195259-05541600-9933-11ea-9261-93c4a60cb962.gif)

1. Touch the map.
2. Open the drawer.
3. Webview becomes unresponsive.
4. Reloading does not fix it.
5. Pressing the "Fix Freeze" button fixes it.

At the end: my English isn't very good, so it's hard to understand. I'm sorry. 😅

**Comment by macoshita** [comment]
P.S.

I've seen the same problem happen with just repeated swipes at https://kenwheeler.github.io/slick/.
Also, I have a self-built carousel library (sorry I can't publish it), and it still has the same problem.

**Comment by cyanglaz** [comment]
@macoshita Did you mean that you still have the issue even applying the fix you provided above?

**Comment by django-wong** [comment]
I encountered the _same or similar_ problem with mapbox-gl-js in WebView. The map gets stuck after a short while I operating (zoom in or zoom out, move the camera) the map.

I have created a [demo app](https://github.com/django-wong/mapbox-gl-in-webview-on-flutter) so you can reproduce the issue.

The demo just works when I switch to `flutter_webview_plugin`
```diff
diff --git a/lib/screens/map/simple-map-view.dart b/lib/screens/map/simple-map-view.dart
index 6f68aa1..5512b5d 100644
--- a/lib/screens/map/simple-map-view.dart
+++ b/lib/screens/map/simple-map-view.dart
@@ -1,12 +1,14 @@
 import 'package:flutter/material.dart';
-import 'package:webview_flutter/webview_flutter.dart';
+import 'package:flutter_webview_plugin/flutter_webview_plugin.dart';
 
 class SimpleMapViewWidget extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
-    return WebView(
-      initialUrl: "http://127.0.0.1:8080/mapbox-gl-demo/index.html",
-      javascriptMode: JavascriptMode.unrestricted,
-    );
+    return WebviewScaffold(url: "http://127.0.0.1:8080/mapbox-gl-demo/index.html");
   }
 }
\ No newline at end of file
```

Flutter 1.17.2 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 5f21edf8b6 (12 hours ago) • 2020-05-28 12:44:12 -0700
Engine • revision b851c71829
Tools • Dart 2.8.3
iOS: 13.5

**Comment by gcemcw4** [comment]
any updates?

**Comment by macoshita** [comment]
@cyanglaz 

> If observing this state and make it automatically rewrite the state, this problem should be fixed.
I've released the app fixed this way. So far no problems have occurred.

However, this is a workaround.
It would be best if the state of ForwardingGestureRecognizer works as expected.

**Comment by cyanglaz** [comment]
@macoshita Thanks for providing the workaround. I have filed a bug/feature request to Apple and see if the WKWebView can be fixed. Otherwise, if the workaround works, it might be our best short term solution. Do you mind to open a PR with the workaround? I'm happy to take a look!

**Comment by jacksonyuanjx** [comment]
> I encountered the _same or similar_ problem with mapbox-gl-js in WebView. The map gets stuck after a short while I operating (zoom in or zoom out, move the camera) the map.
> 
> I have created a [demo app](https://github.com/django-wong/mapbox-gl-in-webview-on-flutter) so you can reproduce the issue.
> 
> The demo just works when I switch to `flutter_webview_plugin`
> 
> ```diff
> diff --git a/lib/screens/map/simple-map-view.dart b/lib/screens/map/simple-map-view.dart
> index 6f68aa1..5512b5d 100644
> --- a/lib/screens/map/simple-map-view.dart
> +++ b/lib/screens/map/simple-map-view.dart
> @@ -1,12 +1,14 @@
>  import 'package:flutter/material.dart';
> -import 'package:webview_flutter/webview_flutter.dart';
> +import 'package:flutter_webview_plugin/flutter_webview_plugin.dart';
>  
>  class SimpleMapViewWidget extends StatelessWidget {
>    @override
>    Widget build(BuildContext context) {
> -    return WebView(
> -      initialUrl: "http://127.0.0.1:8080/mapbox-gl-demo/index.html",
> -      javascriptMode: JavascriptMode.unrestricted,
> -    );
> +    return WebviewScaffold(url: "http://127.0.0.1:8080/mapbox-gl-demo/index.html");
>    }
>  }
> \ No newline at end of file
> ```
> 
> Flutter 1.17.2 • channel stable • https://github.com/flutter/flutter.git
> Framework • revision [5f21edf](https://github.com/flutter/flutter/commit/5f21edf8b66e31a39133177319414395cc5b5f48) (12 hours ago) • 2020-05-28 12:44:12 -0700
> Engine • revision b851c71829
> Tools • Dart 2.8.3
> iOS: 13.5

I can confirm that using the `flutter_webview_plugin` resolved my issue. This will likely be my temporary workaround for now. Hopefully that helps someone!
My issue with `webview_flutter` came from updating my phone to iOS 13.5... initially it was on iOS 12 but the old library seems to work fine.

**Comment by anaimi** [comment]
Applying the workaround from @macoshita is causing a full webview refresh.
flutter_webview_plugin doesn't work for us as we have widgets and alerts over the webview.

Unfortunately this is impacting our iOS users in production. Any tips highly appreciated.

**Comment by paolon** [comment]
> @macoshita Thanks for providing the workaround. I have filed a bug/feature request to Apple and see if the WKWebView can be fixed. Otherwise, if the workaround works, it might be our best short term solution. Do you mind to open a PR with the workaround? I'm happy to take a look!

@cyanglaz Please have news of the bug report? Is it possible to have a link to track the request you made to Apple? With @andrearubeis we have an app in production and we would like to understand if Apple will be able to solve the problem in an upcoming iOS release.

**Comment by cyanglaz** [comment]
@paolon The bug was filed using Apple's Radar system and can be only accessed using the account that filed the bug .

**Comment by qinglongy** [comment]
I'm found this bug in iOS13.4-13.5, in iOS 13.3 is work fine.in Android is work fine.
I‘m open same URL by iOS safari  in iOS13.4-13.5 work fine.
I'm use [_flutter_webview_plugin_] in iOS13.4-13.5 work fine. but this plugin is not in widget tree, so i can't use it.
BTW, my flutter version is 1.17.0.
Sorry, my english is not good, hope can understand.

**Comment by macoshita** [comment]
As a matter of fact, I don't use the webview_flutter plugin.
I use a plugin I created for a specific requirement.
Also, no Refresh Issue is happening in that plugin.
I'm too busy to find out what causes the refresh and create a pull request. Sorry. 🙇

**Comment by Keanyuan** [comment]
After all this time, is there a solution?

**Comment by SAGARSURI** [comment]
Even we are stuck on the same issue. Any progress?

**Comment by lesliechueng1996** [comment]
We are stuck on the same issue too. Any progress?

**Comment by huluwa-dev** [comment]
Same issue here, any solution?

**Comment by v1z4** [comment]
I have the same issue, but I only have it when I have WebView inside on SingleChildScrollView.

**Comment by huoyingjianshen** [comment]
I have the same issue on iOS above 13.4.

The app just works when I switch to flutter_webview_plugin on iOS 13.5.1, which can't set useragent on iOS and display white in some areas after loading on android 10 .

According to the document,on iOS the WebView widget is backed by a WKWebView,but WKWebView should work properly on iOS.

**Comment by cyanglaz** [comment]
The `WKDeferringGestureRecognizer`(Webkit Internal API) in the WKWebView seems to be the one that is blocking the `ForwardingGestureRecognizer` in the flutter engine. 
https://github.com/WebKit/webkit/blame/master/Source/WebKit/UIProcess/ios/WKDeferringGestureRecognizer.mm
This is a new gesture recognizer that was added about 7 months ago.

**Comment by cyanglaz** [comment]
Also, https://github.com/flutter/flutter/issues/30143 might be related

**Comment by evdinursan** [comment]
> 
> 
> Updates:
> The `WKDeferringGestureRecognizer`(Webkit Internal API) in the WKWebView seems to be the one that is blocking the `ForwardingGestureRecognizer` in the flutter engine.
> https://github.com/WebKit/webkit/blame/master/Source/WebKit/UIProcess/ios/WKDeferringGestureRecognizer.mm
> This is a new gesture recognizer that was added about 7 months ago.

Sorry for this question, but how this issue can be fixed? THANK YOU!!!!

**Comment by cyanglaz** [comment]
@evdinursan There are 2 separate issues here:

The issue about `WKDeferrtingGestureRecognizer` blocking the `ForwardingGestureRecognizer` issue can be fixed from flutter engine, which I plan to fix soon. 

Another issue is that `touchStart` event is not blocked by flutter widget (#30143). I'm still trying to find solutions for this issue.

**Comment by evdinursan** [comment]
> @evdinursan There are 2 separate issues here:
> 
> The issue about `WKDeferrtingGestureRecognizer` blocking the `ForwardingGestureRecognizer` issue can be fixed from flutter engine, which I plan to fix soon.
> 
> Another issue is that `touchStart` event is not blocked by flutter widget (#30143). I'm still trying to find solutions for this issue.

I have an application that should go live pretty soon and because of this issue, we will postpone the iOS deployment, so if you need any kind of debugging data, please let ne know.

Thanks

**Comment by marlon2brando** [comment]
> @evdinursan There are 2 separate issues here:
> 
> The issue about `WKDeferrtingGestureRecognizer` blocking the `ForwardingGestureRecognizer` issue can be fixed from flutter engine, which I plan to fix soon.
> 
> Another issue is that `touchStart` event is not blocked by flutter widget (#30143). I'm still trying to find solutions for this issue.



> > @evdinursan There are 2 separate issues here:
> > The issue about `WKDeferrtingGestureRecognizer` blocking the `ForwardingGestureRecognizer` issue can be fixed from flutter engine, which I plan to fix soon.
> > Another issue is that `touchStart` event is not blocked by flutter widget (#30143). I'm still trying to find solutions for this issue.
> 
> I have an application that should go live pretty soon and because of this issue, we will postpone the iOS deployment, so if you need any kind of debugging data, please let ne know.
> 
> Thanks

I stuck the same situation with you!

**Comment by evdinursan** [comment]
Hi everyone! Wanted to let you know that this morning I updated my Xcode to 15.6 and now the issue is gone. If you deploy it on a phisical device (ios) you should be aware that you also should update the iPhone to 13.6 in order to see the fix.

**Comment by jsantaana** [comment]
> Hi everyone! Wanted to let you know that this morning I updated my Xcode to 15.6 and now the issue is gone. If you deploy it on a phisical device (ios) you should be aware that you also should update the iPhone to 13.6 in order to see the fix.
> 
> Still, if you find a fix for the older version, that would be great.

Just tested my app without any changes after updating my iPad to 13.6 and can confirm is working now!! Awesome. Thanks @evdinursan for the good news!

**Comment by ztigunes** [comment]
İ have same issue. Are there any solution except ios 13.6 upgrade ?

**Comment by adnanefouam** [comment]
is there any fix for ios version  under  13.6 ?

**Comment by Hashpurple** [comment]
I have this problem is ios version 13.6 either.
Is there any update?

**Comment by ztigunes** [comment]
Other webview packages  also not working on ios. We had to delay the ios deployment because of this.  
webview_flutter not updated since May 21.  Do you have a road map for this bug? Thanks. Best regards.

**Comment by JohnAtWassa** [comment]
The bug seems to be solved with the latest public beta (7).

**Comment by ztigunes** [comment]
> The bug seems to be solved with the latest iOS 14 public beta (7).

this issue still continue in ios 13.5 and 13.4. webview_flutter not updated since May 21.

**Comment by njovy** [comment]
I am having this issue after update to 1.20.
It isn't affected by aniOS version because the old one that I published to market works fine. Maybe this is a new issue.

**Comment by njovy** [comment]
I updated my Xcode to 11.7 and iOS 13.7 the issue still appears.
I have no idea what to do with this bug.

**Comment by cyanglaz** [comment]
@njovy were you seeing the issue fixed in 13.6 but happened again in 13.7?

**Comment by njovy** [comment]
@cyanglaz Thank you very much for your attention. The webview gesture issue that I found isn't related to a specific iOS version. It's related to flutter SDK versions. The issue I found happens on iOS 13.5, iOS 13.6, 13.7, iOS 11.4.1 so I don't think this issue is iOS version specific.
I also found what version causes this as well. I tested my app with 1.19.0-4.3.pre, 1.20.0, 1.20.1, 1.20.2, 1.20.3, and 1.21.0-9.2.pre. Starting from 1.20.2, my app no longer works as expected. When I launch a screen that has an WebView, I can't touch or scroll inside WebView. Plus, when I swipe right to go back to screen, the screen got stuck and freezes in the middle of swiping.

**Comment by njovy** [comment]
@cyanglaz One thing I noticed is that if I display the problematic WebView screen as an initial screen, it works as expected. It scrolls. The webview screen is an agreement screen showing a privacy and terms, etc so it doesn't make sense to put it as a home screen. I am not sure which patch made this change starting from 1.20.2.

**Comment by njovy** [comment]
![ezgif-2-dcfab5a4f0ba](https://user-images.githubusercontent.com/1792070/92792672-b7cf4d80-f3e8-11ea-9a0c-128b8989113f.gif)
@cyanglaz Above is a screen recording of this issue.
When the last screen is displayed, I tried to scroll but it didn't scroll inside an WebView. When I swipe right to go back, it stuck and the app freezes, I can't touch anything. Images still animating so I wouldn't say the app freezes but every gesture event doesn't work.

**Comment by cyanglaz** [comment]
@njovy I think you have encountered a different issue. Also this issue should be fixed by https://github.com/flutter/engine/pull/20708
If you try flutter on master channel, it should be fixed. 
The fix should also be cherry-picked and you can expect the fix in flutter 1.20.3

**Comment by njovy** [comment]
@cyanglaz No it's not fixed in 1.20.3. That's why I thought this could be a new issue. I already tried 1.20.3, and 1.21.0-9.2.pre. Both sdk versions have this https://github.com/flutter/engine/pull/20708 fix but the issue still appears.

**Comment by ztigunes** [comment]
https://gist.github.com/Voxar/5dbef0b76cc8cb55d2fe74a492d41c54#file-add-a-guardian-for-flutters-touch-interceptor-state-patch that can help. but it does not offer a complete solution. I'm showing a webview and a vis.js timeline in a tabpage. I show loadingindicator by putting webview into stack and I support right-left swipe to switch tabpage. The page only works if there is a webview, but when you use the gesture event from within another widget, the webview freezes. Bug still continue after ios 13.4

**Comment by njovy** [comment]
@cyanglaz Any comments regarding this issue? I can create a new issue if this bug is not related to this.

**Comment by cyanglaz** [comment]
@njovy looks like what you are experiencing is not related to this issue. If you could create a new bug with details explaining the issue, that'd be great!

**Comment by LukasP94** [comment]
![ezgif com-gif-maker](https://user-images.githubusercontent.com/66218871/93197501-01eb6280-f74c-11ea-84b2-0a377a78793a.gif)
 I have similar issue. context Manu is shown even I click flutter button

**Comment by ztigunes** [comment]
> ![ezgif com-gif-maker](https://user-images.githubusercontent.com/66218871/93197501-01eb6280-f74c-11ea-84b2-0a377a78793a.gif)
> I have similar issue. context Manu is shown even I click flutter button


Don't put the flutter button on the map. For example, you can put the map and button in the column. Then you will see that there is no freeze. However, if there are multiple gesture events in the same area, webview also freezes. And this is happening in all ios versions above 13.0.

**Comment by cyanglaz** [comment]
@ztigunes have you tried iOS 13.6 and above?

**Comment by LukasP94** [comment]
I have tried also 13.7

**Comment by ztigunes** [comment]
If we add a webview as a tabview only and swipe left and right in tabs, both the webview freezes and the swipe between tabpages is broken. (use two finger for swipe and try multi touch)  All tabviews freeze. This situation is still happening on ios 13.6 and above.  if you give a webview as TabBarView with initial url 'https://www.google.com/maps' at 'https://flutter.dev/docs/cookbook/design/tabs' this sample, you can see all freeze.

**Comment by cyanglaz** [comment]
@ztigunes What you described could be a different issue. I will give it a try and get back to you.

**Comment by njovy** [comment]
@cyanglaz https://github.com/flutter/engine/pull/20050 This is the only patch that can cause this issue because my webview has been working fine until 1.20.1.

**Comment by cyanglaz** [comment]
@njovy Interesting. It looks like a different issue. Could you please open a new github issue with details(flutter version, reproduction steps, sample code), and tag me there?

**Comment by njovy** [comment]
@cyanglaz https://github.com/flutter/flutter/issues/66044 is the new issue I just opened. As I stated there, I could not find a way to reproduce this issue at the moment. If I find a way, I will add there.

**Comment by ztigunes** [comment]
> @ztigunes What you described could be a different issue. I will give it a try and get back to you.

@cyanglaz, is the #66044 issue related to this?

**Comment by cyanglaz** [comment]
@ztigunes The issue you are facing might be related to #66044, this particular issue is related to WKWebView on iOS 13.4 and 13.5

**Comment by drkhannah** [comment]
this is happening for me as well using flutter 1.20.4 passing a gestureRecognizer to a UIKitView widget

I use the same gesture recognizer for Android and iOS, but it freezes the screen on iOS and works fine on Android, this just started happening in Flutter 1.20.1

SizedBox(
              width: double.infinity,
              height: 130,
              child: (Platform.isIOS)
                  ? UiKitView(
                      gestureRecognizers:
                          <Factory<OneSequenceGestureRecognizer>>[
                        Factory<OneSequenceGestureRecognizer>(
                          () => PlatformViewHorizontalGestureRecognizer(),
                        ),
                      ].toSet(),
                      viewType: 'Goals',
                    )
                  : AndroidView(
                      gestureRecognizers:
                          <Factory<OneSequenceGestureRecognizer>>[
                        Factory<OneSequenceGestureRecognizer>(
                          () => PlatformViewHorizontalGestureRecognizer(),
                        ),
                      ].toSet(),
                      viewType: 'Goals',
                    ),
            )),

**Comment by cyanglaz** [comment]
The issue is due to the WKWebView aggressively blocking the delaying gesture recognizer to block gesture recognizers on the `WKWebView` on iOS 13.4 and 13.5

Unfortunately, after spending major time looking into this bug, I think this is unfixable under the current platform view architecture. The only way that I think we can fix it is to merge the UIThread and PlatformThread in flutter engine to make synchronized gesture blocking. The con of such a fix out weights the pro, especially the issue only occurs in some old iOS versions.

I'm going to close this issue because we don't have a plan to fix it for iOS 13.4 and 13.5.

**Comment by cyanglaz** [comment]
We should also update the documentation of the webview plugin to indicate this limitation of the plugin on iOS 13.4 and 13.5. I opened a new issue to track this limitation. https://github.com/flutter/flutter/issues/67541

**Comment by mehmetf** [comment]
@cyanglaz Do we have proof that this is fixed in later iOS versions like 14.1 and 2? customer:money observed a suspiciously similar problem on iOS 14.2.

**Comment by mehmetf** [comment]
With the repro steps in https://github.com/flutter/flutter/issues/50420, I indeed do not see this problem on iOS 14.3; I asked customer:money to come up with a reproducible use case.

**Comment by pichillilorenzo** [comment]
Tested on a real iPhone 12 Pro and this issue persists on iOS 14.4, using `webview_flutter: ^2.0.0-nullsafety.5`.

After some hard debugging, I found that the problem is somehow related to javascript touch (or whatever) events added to HTML elements.

If you set `javascriptMode: JavascriptMode.disabled` (which is the default value), flutter widgets, such as buttons, that are over the Flutter WebView widget seem to work as expected, and the touch gestures are not propagated to the WebView (WKWebView) under this buttons.

<details>
  <summary>CODE EXAMPLE with JavascriptMode.disabled</summary>

  ```dart
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Example app'),
        ),
        body: Stack(
          children: [
            WebView(
              initialUrl: "https://www.google.com",
              javascriptMode: JavascriptMode.disabled,
            ),
            Container(
              child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.end,
                  children: [
                    ElevatedButton(onPressed: () {
                      print("button 1");
                    }, child: Text("TEST TEST TEST TEST TEST TEST")),
                    ElevatedButton(onPressed: () {
                      print("button 2");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 3");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 4");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 5");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 6");
                    }, child: Text("test")),
                  ]),
            ),

          ],
        ),
      ),
    );
  }
}
  ```
</details>

https://user-images.githubusercontent.com/5956938/108504848-d8180800-72b6-11eb-8f90-319216893e8b.mov

Instead, if you set `javascriptMode: JavascriptMode.unrestricted`, (which is the value used, probably, most of the times), flutter buttons touch gestures are propagated to the Flutter WebView widget, but only on HTML elements that listen to **javascript touch events** (for example: [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) on `touchstart`, `click`, etc.).
So, If you tap on a Flutter Button that is over an HTML element that doesn't listen to any javascript touch events, the tap event doesn't propagate to the Flutter WebView.
Also, I noticed that the propagation is not instant, but only if you tap or long-press two times in the **same area** of the Flutter WebView:
- the first time, you tap or long-press on a flutter button that is over an HTML element with the javascript event listener
- the second time, you can tap or long-press **everywhere** in the same area.

<details>
  <summary>CODE EXAMPLE with JavascriptMode.unrestricted</summary>

  ```dart
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Example app'),
        ),
        body: Stack(
          children: [
            WebView(
              initialUrl: "https://www.google.com",
              javascriptMode: JavascriptMode.unrestricted,
            ),
            Container(
              child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.end,
                  children: [
                    ElevatedButton(onPressed: () {
                      print("button 1");
                    }, child: Text("TEST TEST TEST TEST TEST TEST")),
                    ElevatedButton(onPressed: () {
                      print("button 2");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 3");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 4");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 5");
                    }, child: Text("test")),
                    ElevatedButton(onPressed: () {
                      print("button 6");
                    }, child: Text("test")),
                  ]),
            ),

          ],
        ),
      ),
    );
  }
}
  ```
</details>

https://user-images.githubusercontent.com/5956938/108504740-aacb5a00-72b6-11eb-8c48-89c921ff5830.mov

To summarize: if a flutter widget (such as `ElevatedButton`) is over an HTML element that listens to any of the javascript touch events, the touch gesture (`tap`, `long-press`, ecc.) is propagated to the Flutter WebView, otherwise, it works as expected and the touch gesture is not propagated.

I'm using Flutter from "Channel master, 1.27.0-2.0.pre.43".
<details>
  <summary>flutter doctor</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.27.0-2.0.pre.43, on macOS 11.1 20C69 darwin-x64, locale it-IT)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2020.1.2)
[✓] VS Code (version 1.52.1)
[✓] Connected device (3 available)

• No issues found!
```
</details>

I'm writing here because there are also people having the same problem with my [InAppWebView](https://github.com/pichillilorenzo/flutter_inappwebview) plugin.
Hope it helps to resolve this iOS issue!

**Comment by stellarbeam** [comment]
> If you set `javascriptMode: JavascriptMode.disabled` (which is the default value), flutter widgets, such as buttons, that are over the Flutter WebView widget seem to work as expected, and the touch gestures are not propagated to the WebView (WKWebView) under this buttons.

Indeed. I could reproduce this behaviour. 
I'm facing the same issue on **Android** emulators as well as physical devices, so maybe this affects Android as well.

Output of `flutter doctor` :
```
[✓] Flutter (Channel stable, 2.0.1, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[✓] Android Studio
[✓] VS Code (version 1.54.2)
[✓] Connected device (1 available)

! Doctor found issues in 1 category.
```

**Comment by stellarbeam** [comment]
> I tried adding a GestureDetector on top of WebView, but it does not work as expected.

Okay! So I figured out the mystery. 
Upon observing the gesture area debug logs, I noticed that the inner GestureDetector's (.e. WebView's parent's ) tap recognizer   was indeed receiving the tap gesture, but WebView's tap recognizer would win everytime.

I only had to figure out a way to make sure the GestureDetector _always self-declares a win_ as soon as it receives a tap gesture. 
So instead of using GestureDetector (which in turn uses the in-built tap recognizer), I made use of RawGestureDetector by giving it a custom tap recognizer. 

<details>
<summary>Code for CustomTapRecognizer</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';

class CustomTapRecognizer extends OneSequenceGestureRecognizer {
  final void Function() onTap;
  Offset pointerDownPosition;

  CustomTapRecognizer({@required this.onTap});

  @override
  void addPointer(PointerEvent event) {
    if (event is PointerDownEvent) {
      startTrackingPointer(event.pointer);
      pointerDownPosition = event.position;
    } else {
      stopTrackingPointer(event.pointer);
    }
  }

  @override
  void handleEvent(PointerEvent event) {
    if (event is PointerUpEvent) {
      Offset pointerUpPosition = event.position;
      if (pointerUpPosition - pointerDownPosition == Offset.zero) {
        resolve(GestureDisposition.accepted);
      } else {
        resolve(GestureDisposition.rejected);
      }
      stopTrackingPointer(event.pointer);
    }
  }

  @override
  void acceptGesture(int pointer) {
    super.acceptGesture(pointer);
    onTap();
  }

  @override
  String get debugDescription => "tap";

  @override
  void didStopTrackingLastPointer(int pointer) {}
}
```

</details>

Then I used it as follows:
```dart 
RawGestureDetector( // this as replacement of inner GestureDetector
  gestures: <Type, GestureRecognizerFactory>{
    CustomTapRecognizer:
        GestureRecognizerFactoryWithHandlers<CustomTapRecognizer>(
      () => CustomTapRecognizer(onTap: _onTap),
      (CustomTapRecognizer instance) {},
    )
  },
  child: SizedBox(
    child: VideoBox(_onChangeStatus), // contains WebView
    height: playerHeight,
    width: playerWidth,
  ),
),
```
 New structure: `GestureDetector > RawGestureDetector > WebView`

Now suppose you have another widget over a WebView that needs to recognize a tap, say a button,
then you can create a custom widget for the button (without in-built gesture handlers) and wrap it with RawGestureDetector.

If you need other type of gestures, here are the articles that helped me understand the concepts:
[Gesture disambiguation - Flutter Docs](https://flutter.dev/docs/development/ui/advanced/gestures#gesture-disambiguation)
[Flutter Deep Dive: Gestures](https://medium.com/flutter-community/flutter-deep-dive-gestures-c16203b3434f)
[Combining multiple GestureDetectors in Flutter](https://medium.com/koahealth/combining-multiple-gesturedetectors-in-flutter-26d899d008b2)

**Comment by pichillilorenzo** [comment]
Thanks @stellarbeam but, to me, this seems more a workaround than a solution.
I think that it should already work without implementing that. Just like it would work using Flutter widgets (non-PlatformView).
I think that everyone expects this to work as it should be, that is if I put a button over another widget (that is a `PlatformView` or not, it shouldn't care) the button should win and capture the event (if I didn't specify some custom behaviour).

**Comment by stellarbeam** [comment]
Quoting @amirh , (from [here](https://github.com/flutter/flutter/issues/49084#issuecomment-575812696))

> The **TapGestureRecognizer only accepts gestures as the default winner** (when there are no other recognizers the claim the gesture, Dartdoc).
> 
> A PlatformViewSurface currently always contributes a member to the arena which prevents the tap detector from winning.
> 
> In some sense this is working as intended, as the tap recognizer says "I only recognizer when I know nothing else wants the touches", and a platform view that has non transparent hit test behavior is always "interested" in the touch events.

So I guess this is the behaviour that should be expected as per _gesture arena_ rules. But, I agree, that this current behaviour is not what will be expected by users. 

This behaviour is not a bug per se, because that is what rules say, but a change would definitely be helpful.

**Comment by pichillilorenzo** [comment]
Yeah, I hope this is something feasible for the Flutter team. Thanks again!

**Comment by stellarbeam** [comment]
Mention not! Happy to contribute!

**Comment by Rob-wang** [comment]
I can solve this problem by using flutter_webview_plugin. But it cannot call the display flutter component. If I use UiKitView to wrap it, will there be the same problem in iOS 13.4 and iOS 13.5?

 Looking forward to your reply, it will be of great help to me. thanks
